topology: add block device topology probe

trgill opened the pull request:
The probe reads JSON from the lsblk command to get a
description of the block devices on the system.  The
JSON is parsed into structures that describe the relationships
between the devices (parent/child,LVM,MD,multipath).

If the environment variable SKYDIVE_BLOCKDEV_TEST_FILE is set
the file is read rather than invoking "lsblk --json".  This is
for testing more complex block device setups that might not
be available on the local system.

Signed-off-by: Todd Gill <tgill@redhat.com>

trgill commented:
This is a WIP - I opened the PR to give people an early look to see if it is headed in the right direction.

To test the probe - set SKYDIVE_BLOCKDEV_TEST_FILE to a file containing the following JSON:

https://pastebin.com/bJfcFxin

The JSON has the contents of "lsblk" from a system with several types of block storage.

hunchback commented on topology/probes/blockdev/metadata.go:
```diff
@@ -0,0 +1,98 @@
+	"github.com/skydive-project/skydive/graffiti/graph"
+)
+
+// Metadata describe the metadata of a docker container
```
s/docker container/block device/

trgill replied:
thanks.

lebauce commented on agent/topology_probes.go:
```diff
@@ -47,6 +48,8 @@ import (
 func NewTopologyProbe(name string, ctx tp.Context, bundle *probe.Bundle) (probe.Handler, error) {
 	switch name {
+	case "blockdev":
+		return new(blockdev.ProbeHandler).Init(ctx, bundle)
```
We recently merged a PR so that Skydive handles probes as external .so
Therefore, we now try to use the `module.NewProbe(ctx, bundle)` scheme. Could you please do this small refactoring ?

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,462 @@
+	p.addGroupByName(blockdevGroupName, "")
+
+	if _, err := topology.AddLink(p.Ctx.Graph, p.Ctx.RootNode, p.Groups[blockdevGroupName], "connected", nil); err != nil {
+		p.Ctx.Logger.Error(err)
```
Should we return err ?

trgill replied:
yes.  will fix

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,462 @@
+}
+
+// Start the probe
+func (p *ProbeHandler) Start() error {
```
You could make NewProbe (renamed from Init) return a probe wrapper as in https://github.com/skydive-project/skydive/blob/master/topology/probes/docker/docker.go#L281
Therefore, you don't need to implement Start and Stop, but only the Do method. That would give something like :

```
 func (p *ProbeHandler) Do(ctx context.Context, wg *sync.WaitGroup) error {
    p.addGroupByName(blockdevGroupName, "")
    ...

    wg.Add(1)
    go func() {
        defer wg.Done()

        after := time.After(30*time.Second)
        for {
            if err := p.connect(); err != nil {
                p.Ctx.Logger.Error(err)
                return
            }

            select {
                case <- ctx.Done():
                    return
                case <-after:
            }
        }
    }
```

trgill replied:
Ok - thanks.

lebauce commented on topology/probes/blockdev/metadata.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+// MetadataDecoder implements a json message raw decoder
+func MetadataDecoder(raw json.RawMessage) (common.Getter, error) {
```
You need to register your MetadataDecoder to the analyzer in the `init` function of analyzer/server.go

trgill replied:
ok

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,462 @@
+		"WWN":        WWN,
+		"Manager":    managerType,
+		"BlockID":    p.getID(blockdev),
+		"MetaData":   blockdevMetadata,
```
We are already in metadata, so a MetaData field is a bit confusing.
What about `BlockDevice` or `BlockDev` ? By the way, you should use the same as in the `init` function of analyzer/server.go

I would also suggest to move "BlockID", "WWN" and "MajorMinor" to the Metadata struct you added

trgill replied:
ok

lebauce commented on topology/probes/blockdev/metadata.go:
```diff
@@ -0,0 +1,98 @@
+//go:generate go run github.com/skydive-project/skydive/scripts/gendecoder -package github.com/skydive-project/skydive/topology/probes/blockdev
```
Please add the metadata_easyjson.go and metadata_gendecoder.go files to the repository

trgill replied:
ok

lebauce commented on topology/probes/blockdev/metadata.go:
```diff
@@ -0,0 +1,98 @@
+//go:generate go run github.com/skydive-project/skydive/scripts/gendecoder -package github.com/skydive-project/skydive/topology/probes/blockdev
+//go:generate go run github.com/safchain/easyjson/easyjson $GOFILE
```
go run github.com/mailru/easyjson/easyjson

lebauce commented on topology/probes/blockdev/metadata.go:
```diff
@@ -0,0 +1,98 @@
+type Metadata struct {
+	ID           string
+	Name         string
+	Alignment    int
```
We usually try to use int64 instead of int in Metadata

lebauce commented on topology/probes/blockdev/metadata.go:
```diff
@@ -0,0 +1,98 @@
+func MetadataDecoder(raw json.RawMessage) (common.Getter, error) {
+	var m Metadata
+	if err := json.Unmarshal(raw, &m); err != nil {
+		return nil, fmt.Errorf("unable to unmarshal docker metadata %s: %s", string(raw), err)
```
s/docker/blockdev/

lebauce commented on topology/probes/blockdev/no_blockdev.go:
```diff
@@ -0,0 +1,43 @@
+	tp "github.com/skydive-project/skydive/topology/probes"
+)
+
+// ProbeHandler describes a Docker topology graph that enhance the graph
```
s/Docker/Block device/

lebauce commented:
run tests

lebauce commented:
run tests

safchain commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,462 @@
+	Pkname       string `json:"pkname"`
+	Pttype       string `json:"pttype"`
+	Ptuuid       string `json:"ptuuid"`
+	Ra           int    `json:"ra"`
```
It seems that depending of the lsblk version the output types are not the same :( So you would have to probably maintain mutliple struct. With my version I get a string type for the Ra field
lsblk from util-linux 2.32.1

lebauce replied:
Or just comment `Ra` as it's not used

lebauce replied:
An other option would be to json.Unmarshal to get an interface{}
Then use mapstructure to map it to a BlockDevice struct (we already use mapstructure in a couple of places)

trgill replied:
It is the "Read Ahead" for the device.  I was testing with 2.33.2 and it is an integer in that version.  I'll look at this more.

trgill replied:
It looks like lsblk version 2.33 "has been improved to differentiate between numbers,
booleans and strings in JSON output".  See:

https://mirrors.edge.kernel.org/pub/linux/utils/util-linux/v2.33/v2.33-ReleaseNotes

I added some code to deal with integers and boolean types possibly encoded as strings.  But, I think it introduces a problem with how skydive looks at the meta data - I'm not sure what the issue is, but currently the blockdev meta data doesn't expand in the UI.  (?)

I'll see if I can make it work.

safchain commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,462 @@
+	}
+
+	jsonBlockdevs := string(cmdOut[:])
+	if err = json.Unmarshal([]byte(jsonBlockdevs), &result); err != nil {
```
json.Unmarshal(cmdOut, ....
is not working ? instead of passing via a string

lebauce replied:
if err = json.Unmarshal([]byte(cmdOut[:]), &result); err != nil {

trgill replied:
Ok.

safchain commented:
Hi Todd, thanks for submitting this PR. Just to let you know that we are actively working on a new WebUI. Here how your PR looks like for now. There is a way to configure some rendering things but I did not do anything for now.

![screenshot-0 0 0 0_8080-2019 10 10-14_17_00](https://user-images.githubusercontent.com/2663912/66568150-14871000-eb69-11e9-87c5-590d46664861.png)

trgill commented:
@safchain the new UI looks great!

The JSON was generated on a system that has a mirrored LVM logical volume that uses 2 multipath devices (4 paths each).  It might be a good test case for the new UI.  This is what it looks like currently:
![skydive-lvm-over-multipath](https://user-images.githubusercontent.com/13203542/66577803-f8bf4200-eb47-11e9-9783-8b5005b3b81e.png)

trgill commented:
@hunchback @lebauce @safchain thanks for the feedback.  I'll make the suggested changes.

The icons are intended as just place holders for something better.  I think it might be useful to visually distinguish physical block devices, DM, LVM, MD, and multipath etc.

I started looking at using "iostat" to map in performance data to the physical devices.   This is a JSON example of the multipath devices under and IO load:

https://pastebin.com/NT1YpBE5

Do you have any example probes that I should follow to map in the performance data?

lebauce commented:
run skydive-scale-tests

lebauce commented on statics/js/components/topology.js:
```diff
@@ -1031,8 +1031,8 @@ Node.prototype = {
   isCaptureAllowed: function() {
-    var allowedTypes = ["device", "veth", "ovsbridge", "geneve", "vlan", "bond", "ovsport",
-                        "internal", "tun", "bridge", "vxlan", "gre", "gretap", "dpdkport"];
+    var allowedTypes = ["device", "veth", "blockdev", "blockdevleaf", "blockdevlvm", "ovsbridge", "geneve",
```
I'm not sure you need this

trgill replied:
looks like things work fine without - so I removed from PR

trgill commented on tests/blockdev_test.go:
```diff
@@ -0,0 +1,324 @@
+		return err
+	}
+
+	// TODO restart the blockdev probe?
```
Is there a way to restart the probe here?  Or should I take a different approach?

lebauce commented on tests/tests.go:
```diff
@@ -80,7 +80,8 @@ agent:
   topology:
     probes:
-    - ovsdb
+	- ovsdb
```
Please fix indentation (spaces instead of tabs)

trgill replied:
oops.  will fix.

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+			groupNode = p.registerBlockdev(blockdev.Children[i], blockdev.WWN)
+		}
+		if blockdev.Type == multipathType {
+			groupNode = p.addGroup(blockdev, parentWWN)
```
addGroup seems to require the lock on ProbeHandler (line 178). For simplicity, you could do
```
p.Lock()
defer p.Unlock()
```
at the start of the function, like in `unregisterBlockdev`

trgill replied:
ok

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+		return 0
+	}
+	if utf8.Valid(field) {
+		var i int64
```
nit:
```
if i, err := strconv.ParseInt(strings.Replace(string(field), "\"", "", -1), 10, 64); err == nil {
    return i
}
```

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+}
+
+func (p *ProbeHandler) getBool(field json.RawMessage) bool {
+
```
nit: remove empty line

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+		Ctx:         ctx,
+	}
+
+	p.Ctx = ctx
```
this line can be removed

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+	}
+
+	p.Ctx = ctx
+	p.blockdevMap = make(map[string]blockdevInfo)
```
what about putting this in the `p` initialization above :
```
    blockdevMap: make(map[string]blockdevInfo),
    Groups: make(map[string]*graph.Node),
```

trgill replied:
thanks - that's better

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+type BlockDevice struct {
+	Children []BlockDevice `json:"children"`
+
+	Alignment    json.RawMessage `json:"alignment"` //int64
```
An alternative would be :
- use json.Unmarshal to a map[string]interface{}
- then use https://godoc.org/github.com/mitchellh/mapstructure#WeakDecode to map to a `BlockDevice` struct. It converts strings to ints (and more) so it could handle a change of type of a member.

It's really up to you :-) We can merge it without it

trgill replied:
I think I've implemented it the way you've suggested.  It looks better to me - please have a look and let me know if I got it right. :-)  Will push soon.

trgill replied:
Ok - I see your 2nd commit looks better than what I came up with... I'll cherry-pick and test.

trgill replied:
@lebauce I squashed your JSON parser changes (hope that is ok) and cherry-picked the "lsblk" path changes.

where are you thinking we should put the fake lsblk?

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+	Groups      map[string]*graph.Node
+}
+
+func (p *ProbeHandler) getPath(blockdev BlockDevice) string {
```
nit: what about making `getPath` a method on BlockDevice instead of ProbeHandler

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+	return blockdev.Kname
+}
+
+func (p *ProbeHandler) getID(blockdev BlockDevice) string {
```
same than above

trgill replied:
I agree - but the functions are no longer needed when I the JSON parsing technique

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+	return blockdev.Name
+}
+
+func (p *ProbeHandler) getName(blockdev BlockDevice) string {
```
same than above

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+
+	if testFile != "" {
+		if cmdOut, err = exec.Command("cat", testFile).Output(); err != nil {
+			p.Ctx.Logger.Error(err)
```
No need to log (it will be done by the called), just `return err`

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+		}
+	} else {
+		if cmdOut, err = exec.Command("lsblk", "-pO", "--json").Output(); err != nil {
+			p.Ctx.Logger.Error(err)
```
Same then above

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,515 @@
+	}
+
+	if err = json.Unmarshal([]byte(cmdOut[:]), &result); err != nil {
+		p.Ctx.Logger.Error(err)
```
Same than above

lebauce commented:
To test, I see 2 solutions :
- execute real commands (using losetup ?) but, as I'm not a storage expert, I don't know if this is feasible
- add a configuration knob to be able to specify the location of the `lsblk` executable, so that we can create a fake `lsblk` that returns the JSON output

I created a branch based on yours with 2 additional commits:
- one (https://github.com/lebauce/skydive/commit/a9e2d52d6bcc314d50aa3859942e7fec03f6a848) that implements the second solution and improves testing a bit
- an other (https://github.com/lebauce/skydive/commit/8500b5f7e93e5864bb46e8abfed0870cf2f8a277) that makes use of mapstructure (this has to be checked though)

Feel free to cherry pick or ignore any of them :-)

trgill commented:
@lebauce loopback devices would be an option - but I've had trouble with them when test frameworks fail.  The tear down can be tricky when they are used as storage for higher levels. 

I'm thinking if we could add a few .json files from complex storage environments - we could avoid having to deal with actually using DM/MD/LVM.

lebauce commented on config/config.go:
```diff
@@ -47,6 +47,7 @@ var (
 		"agent.flow.sflow.bind_address":        {"sflow.bind_address"},
 		"agent.flow.sflow.port_min":            {"sflow.port_min"},
 		"agent.flow.sflow.port_max":            {"sflow.port_max"},
+		"agent.topology.blockdev.lsblk_path":   {"lsblk.path"},
```
relocationMap is for legacy configuraion (renamed or moved entries)

trgill replied:
ok - will remove

lebauce commented on tests/tests.go:
```diff
@@ -81,12 +81,15 @@ agent:
     - socketinfo
     - libvirt{{block "agentProbes" .}}{{"\n"}}{{range .AgentProbes}}{{println "    -" .}}{{end}}{{end}}
+    blockdev:
+      lsblk_path:
```
This seems unnecessary (as the SKYDIVE_AGENT_TOPOLOGY_BLOCKDEV_LSBLK_PATH overrides it)

lebauce commented on tests/blockdev_test.go:
```diff
@@ -0,0 +1,334 @@
+				return err
+			}
+
+			if len(nodes) > 0 {
```
len(nodes) == 0 ?

trgill replied:
I was missing:

setupFunction: setupBlockDev

I updated the test.

trgill commented:
When I click on a block device in the UI - the right side displays the top level data, but the BlockDev doesn't expand as it does for other probes.  Is it obvious to anyone what I'm doing wrong?

lebauce commented:
@trgill That was a bug in the UI. We just merged a fix for it. Can you please try to rebase on master ?

trgill commented:
@lebauce thanks!  Rebase over master fixed the problem.

When I expand the metadata, I see minor formatting issues for Label & Mode.  Is this due to the probe adding metadata that includes HTML characters?

![image](https://user-images.githubusercontent.com/13203542/68491293-8f028880-0218-11ea-91ce-52e717972d02.png)

trgill commented:
@safchain I think this is close to displaying basic metrics via iostat.  The table displays properly - but I'm missing some part that populates the data for the table.  Do you see what I'm doing wrong?

trgill commented:
@safchain I think I've at least partially figured out the problem.  I had used the BlockdevMetric JSON annotations to parse the output from iostat.  But, that looks like it causes problems with the UI.  I'll fix and re-push.

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
 	return m1, m2
 }
+
+// BlockdevMetric the blockdevice IO stats
```
Do you think that this struct could be reused by another probe ? if so we can let it here, if not I would move it in the blockdev folder

trgill replied:
OK.  I moved it to blockdev.go.  I ran into some problems with the build if I put it in a separate file.

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+
+// GetBlockdevFields returns the JSON annotation for each field in
+// BlockdevMetric
+func GetBlockdevFields() []string {
```
this should be generated by the gendecoder

trgill replied:
I was thinking the columns displayed in the UI should match the output of "iostat".  So, I call this function in server.go to use these field names rather than what is returned by .GetFieldKeys().

Some of the field names are not allowed in go structures.

safchain replied:
I not sure that it is good the keep the orignal field name as this is not how the field are reported in skydive in general and can/will raise error with the datastore as not all characters are supported. If we want to keep the original information present in the UI then I would prefer to have a translation there ?

trgill replied:
Ok - I changed to follow the conventions used for the other probes. :)

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// IsZero returns true if all the values are equal to zero
+func (im *BlockdevMetric) IsZero() bool {
```
If don't plan to use gremlin to do operation on metrics this function is useless

trgill replied:
Thanks - I removed it.

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// Split splits a metric into two parts
+func (im *BlockdevMetric) Split(cut int64) (common.Metric, common.Metric) {
```
same here

trgill replied:
removed

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// GetStart returns start time
+func (im *BlockdevMetric) GetStart() int64 {
```
same

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// SetStart set start time
+func (im *BlockdevMetric) SetStart(start int64) {
```
same

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// GetLast returns last time
+func (im *BlockdevMetric) GetLast() int64 {
```
same

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// SetLast set last tome
+func (im *BlockdevMetric) SetLast(last int64) {
```
same

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// Add sum two metrics and return a new Metrics object
+func (im *BlockdevMetric) Add(m common.Metric) common.Metric {
```
same

trgill replied:
removed

safchain commented on topology/metrics.go:
```diff
@@ -232,3 +232,190 @@ func (im *InterfaceMetric) Split(cut int64) (common.Metric, common.Metric) {
+}
+
+// Sub subtracts two metrics and return a new metrics object
+func (im *BlockdevMetric) Sub(m common.Metric) common.Metric {
```
same

trgill replied:
removed

safchain commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,620 @@
+		result Devices
+	)
+
+	lsblkPath := p.Ctx.Config.GetString("agent.topology.blockdev.lsblk_path")
```
I think you should check that this is not empty, I think it is possible to set an empty value in the config

trgill replied:
Ok.

safchain commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,620 @@
+		err    error
+		stats  sysstat
+	)
+	if cmdOut, err = exec.Command("/usr/bin/iostat", "-dx", "-o", "JSON", blockdevPath).Output(); err != nil {
```
does it make sense to make it part of the config ?

trgill replied:
yes - done.

safchain commented on scripts/gendecoder/main.go:
```diff
@@ -95,6 +95,8 @@ func handleBasic(getter *getter, name, kind string) {
 	switch kind {
 	case "int", "int8", "int16", "int32", "int64", "uint", "uint8", "uint16", "uint32", "uint64":
 		getter.Fields = append(getter.Fields, field{Name: name, Type: "int64"})
+	case "float32", "float64":
```
unfortunately this is not enough to have float handled by the gendecoder. The function used to retrieve the value are only generated for int64 and I think that the gremlin traversal thing will have to be updated too. Any chance to have int64 only values ? conversion into ms ?

trgill replied:
Ok.  I changed to int64 - it looks like it is just truncating the floating point. I think for a first version that will be ok.

I removed the "LastUpdate" code.  It doesn't really make sense for rates (like KB per second).  I think we can come back to that at some point.

I'd like to be able to display some IO rates on the edges - but it looks like a fair amount of work for the legacy UI.  So, if you support it in the new UI - please keep in mind that a block device link would want to display the rates as IOPs or KB/s etc.

Does the way I implemented the "metrics" make sense?  I didn't see a better way.  The two structures are necessary for parsing the output of iostat and the other for passing to the UI. (?)

safchain replied:
Thanks for changing to int64, we discussed about this limitation during our last public weekly call, we will remove this limitation in the future. I plan to work soon on the link labels on the "new" UI, I'll take care of this. Having two structures one for parsing and another one for the output seems to be a good approach (IMHO).

safchain commented:
Just a comment regarding the output format, otherwise looks good. Test needs to be fixed

trgill commented:
@safchain I think I've addressed your concerns.  Does this look like a decent "first version"?   I'm thinking there are still a lot of improvements to be made - but getting feedback from users might be a good idea.

safchain commented:
@trgill yes pretty decent version. If you can fix functional and unit test + fmt, I'll merge it with pleasure

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,623 @@
+		if node == nil {
+			continue
+		}
+		path, err := node.GetField("Path")
```
This should be done under graph lock.

nit: you can probably use `GetFieldString` (and therefore "%s" in the Sprintf below).

lebauce commented on topology/probes/blockdev/blockdev.go:
```diff
@@ -0,0 +1,623 @@
+}
+
+func (p *ProbeHandler) deleteIfRemoved(currentDevInfo blockdevInfo, newDevInfo []BlockDevice) {
+	blockID, err := currentDevInfo.Node.GetFieldString("BlockID")
```
It seemsBlockID is never set. Should it be "Index" instead ?

lebauce commented:
@trgill I took the liberty to fix the tests that failed by rebasing your PR. I added one commit (https://github.com/skydive-project/skydive/pull/2057/commits/d36956fac0febf928b221101c7506d199edf3b74) to fix the "BlockID" issue. Does it look correct to you ?

lebauce commented:
run functional-tests-backend-elasticsearch

trgill commented:
@lebauce yes - that is correct.  Thanks for the improvement.

safchain commented:
Thanks a lot @trgill
